src/passes/Unsubtyping.cpp

```diff
@@ -512,6 +514,42 @@ struct Unsubtyping : Pass {
     processCasts(sub, super, oldSuper);
   }
 
+  void processDescribed(HeapType sub, HeapType mid, HeapType super) {
+    // We are establishing sub <: mid <: super. If super describes the immediate
+    // supertype of the type sub describes. Then once we insert mid between them
+    // we would have this:
+    //
+    // A -> super
+    // ^     ^
+    // |    mid
+    // |     ^
+    // C -> sub
+    //
+    // This violates the requirement that the descriptor of C's immediate
+    // supertype must be the immediate supertype of C's descriptor. To fix it,
+    // we have to find the type B that mid describes and insert it between A and
+    // C:
+    //
+    // A -> super
+    // ^     ^
+    // B -> mid
+    // ^     ^
+    // C -> sub
+    //
+    // We do this eagerly before we establish sub <: mid <: super so that if
+    // establishing that subtyping requires recursively establishing other
+    // subtypings, we can depend on the invariant that the described types are
+    // always set up correctly beforehand.
+    auto subDescribed = sub.getDescribedType();
+    auto superDescribed = super.getDescribedType();
+    if (subDescribed && superDescribed &&
+        types.getSupertype(*subDescribed) == superDescribed) {
+      auto midDescribed = mid.getDescribedType();
+      assert(midDescribed);
+      process(*subDescribed, *midDescribed);
```

Comment:
This puts `B===midDescribed` in between `A` and `C`, but how does it achieve the property of being the *immediate* supertype? (with nothing in between)

Reply:
If `B` would not be an immedate child of A and immediate parent of C once inserted, then we must already have some other type B' getting in the way and `processDescribed` will not do anything. We have either this:

```
A -> super
^
B'-> mid'

B -> mid

C -> sub
```
or this:
```
A -> super

B -> mid

B'-> mid'
^
C -> sub
```

Once all the implications established after the call to `processDescribed` and on the worklist are processed, the first turns into this:

```
A -> super
^     ^
B'-> mid'
      ^
B -> mid
      ^
C -> sub
```
and the second into this:
```
A -> super
      ^
B -> mid
      ^
B'-> mid'
^     ^
C -> sub
```

Both are perfectly valid.